crates/rust-analyzer/src/cli/diagnostics.rs

```diff
@@ -0,0 +1,74 @@
+//! Analyze all files in project for diagnostics. Exits with a non-zero status
+//! code if any errors are found.
+
+use anyhow::anyhow;
+use ra_db::{SourceDatabaseExt, SourceRootId};
+use ra_ide::{Analysis, Severity};
+use std::{collections::HashSet, path::Path};
+
+use crate::cli::{load_cargo::load_cargo, Result};
+use hir::{db::HirDatabase, Crate, Module};
+
+pub fn diagnostics(path: &Path, load_output_dirs: bool) -> Result<()> {
+    let (host, roots) = load_cargo(path, load_output_dirs)?;
+    let db = host.raw_database();
+    let analysis = host.analysis();
+    let members = roots
+        .into_iter()
+        .filter_map(
+            |(source_root_id, project_root)| {
+                if project_root.is_member() {
+                    Some(source_root_id)
+                } else {
+                    None
+                }
+            },
+        )
+        .collect::<HashSet<_>>();
+
+    let mut found_error = false;
+    let mut visited_modules = HashSet::new();
+    for krate in Crate::all(db) {
+        let module = krate.root_module(db).expect("crate without root module");
+        check_module(module, db, &mut visited_modules, &members, &analysis, &mut found_error);
+    }
+
+    println!();
+    println!("diagnostic scan complete");
+
+    if found_error {
+        println!();
+        Err(anyhow!("diagnostic error detected"))
+    } else {
+        Ok(())
+    }
+}
+
+fn check_module(
+    module: Module,
+    db: &(impl HirDatabase + SourceDatabaseExt),
+    visited_modules: &mut HashSet<Module>,
+    members: &HashSet<SourceRootId>,
+    analysis: &Analysis,
+    found_error: &mut bool,
+) {
+    let file_id = module.definition_source(db).file_id.original_file(db);
+    if !visited_modules.contains(&module) {
+        if members.contains(&db.file_source_root(file_id)) {
+            println!("processing: {}", db.file_relative_path(file_id));
+            for diagnostic in analysis.diagnostics(file_id).unwrap() {
```

Comment:
I'm not sure about walking the module tree and then going back to files. `SourceRoot` has a `walk` method that lists all the files, so you could try using that directly? (you get the `SourceRoot` through `db.source_root(id)`)

Reply:
If I'm remembering correctly, that was my first attempt. The issue was that it included all files rather than just files in the module tree, which meant for rust analyzer itself that it included all the `test-data` for the parser.

Is there a different way to filter those out?

Reply:
Hmm one way might be to use `Semantics::to_module_def(file_id).is_some()`, though I don't know if that's really better than this approach at that point :thinking:

Reply:
Oh, and I tried one other way which was using `Module::diagnostics`, but this has an interface which is not quite as nice (requiring passing in a `DiagnosticSink`, and it doesn't provide the information about the level of the diagnostic (`Error` vs `Warning`). 

As an aside, it seems the diagnostic level (`Error` vs `Warning`) would be part of the `Diagnostic` trait. Is there some reason it isn't?

But I'll try your suggestion of `Semantics::to_module_def(file_id).is_some()`, because it may help solve the "how to print the crate name" problem. Right now, since I am iterating over `hir::code_model::Crate::all` I don't think I have a nice way to get to the `db::input::CrateData`.

Reply:
`Analysis::diagnostics` adds the 'ide-level' concerns for diagnostics, which includes quick fixes and apparently also the severity (I guess that's debatable, but apparently that's how it is right now). I do think it makes sense to use that for this, since it also adds syntax errors and we might want to print fixes (or at least whether they exist) as well.

Reply:
Maybe we even shouldn't filter out the test data files by default, but provide an option to include/exclude globs -- it seems legitimate to get the diagnostics for any file in the same way we would see them in the editor :thinking:

Reply:
> Maybe we even shouldn't filter out the test data files by default, but provide an option to include/exclude globs -- it seems legitimate to get the diagnostics for any file in the same way we would see them in the editor thinking

I'd really like to be able to use this as at tool for avoiding false positives when developing diagnostics. For that use case, I think automatic filtering of files makes sense, although I suppose someone could potentially have a slightly different use case where they'd like to see all diagnostics. My goal here was to create something similar to `cargo check`. But I've created a CLI option to get the behavior you are describing.